### PR TITLE
Add ReadNone attr for Builtin functions

### PR DESCRIPTION
**Pull request by fadeeval**
All constant built in functions should have readnone attribute in LLVM IR. 

Signed-off-by: Aleksander Fadeev <aleksander.fadeev@intel.com>

### CODE REVIEW

**Comment by svenvh** [comment]
Shouldn't builtin functions already have a similar attribute in SPIR-V?

(we might still need to fix #281)

**Comment by fadeeval** [comment]
@svenvh, builtin functions have not definition in SPIRV, so there are no place to specify such similar attribute in SPIRV code, so this PR serves to right translation back to LLVM IR.

**Comment by AlexeySotkin** [comment]
@fadeeval please add a lit test.

File: lib/SPIRV/SPIRVReader.h [hunk 1]
```diff
@@ -220,6 +220,44 @@ class SPIRVToLLVM {
   // OpenCL function always has NoUnwind attribute.
   // Change this if it is no longer true.
   bool isFuncNoUnwind() const { return true; }
+
+  bool isFuncReadNone(const std::string& name) const {
+    if (BuiltInConstFunc.find(name) != BuiltInFunc.end())
+      return true;
+    else
+      return false;
+  }
```
Comment (AlexeySotkin) [inline lib/SPIRV/SPIRVReader.h:229]:
```suggestion
    return BuiltInConstFunc.count(name);
  }
```

File: lib/SPIRV/SPIRVReader.h [hunk 2]
```diff
@@ -220,6 +220,44 @@ class SPIRVToLLVM {
...
+
+  std::set<std::string> BuiltInConstFunc{
```
Comment (AlexeySotkin) [inline lib/SPIRV/SPIRVReader.h:231]: Is it possible to make it `static const` and move it to SPIRVReader.cpp?

Probably it's worth trying to use `llvm::StringSet` here. It has more efficient memory handling. Though I'm not sure it keeps the elements ordered.

File: lib/SPIRV/SPIRVReader.h [hunk 3]
```diff
@@ -45,6 +45,7 @@
 
 #include "llvm/ADT/DenseMap.h"
 #include "llvm/IR/GlobalValue.h" // llvm::GlobalValue::LinkageTypes
+#include "llvm/ADT/StringSet.h"
```
Comment (AlexeySotkin) [inline lib/SPIRV/SPIRVReader.h:48]: Please move it one line up. Let's keep include directives sorted alphabetically

File: test/transcoding/builtin_function_readnone_attr.ll [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  store double %call, double* %b, align 8, !tbaa !7
+  %1 = load double, double* %a, align 8, !tbaa !7
+  %call1 = tail call double @_Z3cosd(double %1) #2
+  store double %call1, double* %b, align 8, !tbaa !7
+  ret void
+}
+
+; CHECK-LLVM: ; Function Attrs: nounwind readnone
+; Function Attrs: convergent nounwind readnone
+declare dso_local double @_Z3expd(double) local_unnamed_addr #1
```
Comment (AlexeySotkin) [inline test/transcoding/builtin_function_readnone_attr.ll:23]: It is not reliable to check comments. I think it should be like
```suggestion
; Function Attrs: convergent nounwind readnone
CHECK: @_Z3expd{{.*}}#[[#Attrs:]]
declare dso_local double @_Z3expd(double) local_unnamed_addr #1
```

File: test/transcoding/builtin_function_readnone_attr.ll [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+; Function Attrs: convergent nounwind readnone
+declare dso_local double @_Z3cosd(double) local_unnamed_addr #1
+
+attributes #0 = { convergent nofree norecurse nounwind uwtable "correctly-rounded-divide-sqrt-fp-math"="false" "disable-tail-calls"="false" "frame-pointer"="all" "less-precise-fpmad"="false" "min-legal-vector-width"="0" "no-infs-fp-math"="false" "no-jump-tables"="false" "no-nans-fp-math"="false" "no-signed-zeros-fp-math"="false" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" "uniform-work-group-size"="false" "unsafe-fp-math"="false" "use-soft-float"="false" }
+attributes #1 = { convergent nounwind readnone "correctly-rounded-divide-sqrt-fp-math"="false" "disable-tail-calls"="false" "frame-pointer"="all" "less-precise-fpmad"="false" "no-infs-fp-math"="false" "no-nans-fp-math"="false" "no-signed-zeros-fp-math"="false" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" "unsafe-fp-math"="false" "use-soft-float"="false" }
```
Comment (AlexeySotkin) [inline test/transcoding/builtin_function_readnone_attr.ll:30]:
```suggestion
CHECK-LLVM: attributes #[[#Attrs]] {{.*}} readnone 
attributes #1 = { convergent nounwind readnone "correctly-rounded-divide-sqrt-fp-math"="false" "disable-tail-calls"="false" "frame-pointer"="all" "less-precise-fpmad"="false" "no-infs-fp-math"="false" "no-nans-fp-math"="false" "no-signed-zeros-fp-math"="false" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" "unsafe-fp-math"="false" "use-soft-float"="false" }
```

File: test/transcoding/builtin_function_readnone_attr.ll [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+  ret void
+}
+
+; CHECK-LLVM: ; Function Attrs: nounwind readnone
+; Function Attrs: convergent nounwind readnone
+declare dso_local double @_Z3expd(double) local_unnamed_addr #1
+
+; CHECK-LLVM: ; Function Attrs: nounwind readnone
+; Function Attrs: convergent nounwind readnone
+declare dso_local double @_Z3cosd(double) local_unnamed_addr #1
```
Comment (AlexeySotkin) [inline test/transcoding/builtin_function_readnone_attr.ll:27]:
```suggestion
; Function Attrs: convergent nounwind readnone
; CHECK-LLVM: @_Z3cosd{{.*}} [[#Attrs]]
declare dso_local double @_Z3cosd(double) local_unnamed_addr #1
```

File: test/transcoding/builtin_function_readnone_attr.ll [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  store double %call1, double* %b, align 8, !tbaa !7
+  ret void
+}
+
+; Function Attrs: convergent nounwind readnone
+; CHECK-LLVM: @_Z3expd{{.*}}#[[#Attrs:]]
+declare dso_local double @_Z3expd(double) local_unnamed_addr #1
+
+; Function Attrs: convergent nounwind readnone
+; CHECK-LLVM: @_Z3cosd{{.*}} [[#Attrs]]
```
Comment (AlexeySotkin) [inline test/transcoding/builtin_function_readnone_attr.ll:26]:
```suggestion
; CHECK-LLVM: @_Z3cosd{{.*}} #[[#Attrs]]
```
It makes sense to run lit tests locally before pushing to GitHub.

File: lib/SPIRV/SPIRVReader.h [hunk 8]
```diff
@@ -220,6 +221,11 @@ class SPIRVToLLVM {
   // OpenCL function always has NoUnwind attribute.
   // Change this if it is no longer true.
   bool isFuncNoUnwind() const { return true; }
+
+  bool isFuncReadNone(const std::string &name) const {
+    return BuiltInConstFunc.count(name);
```
Comment (AlexeySotkin) [inline lib/SPIRV/SPIRVReader.h:226]:
```suggestion
  bool isFuncReadNone(const std::string &Name) const {
    return BuiltInConstFunc.count(Name);
```

File: lib/SPIRV/SPIRVReader.cpp [hunk 9]
```diff
@@ -4426,3 +4428,32 @@ bool llvm::getSpecConstInfo(std::istream &IS,
   }
   return !IS.fail();
 }
+
```
Comment (AlexeySotkin) [inline lib/SPIRV/SPIRVReader.cpp:4431]:
```suggestion

// clang-format off
```

File: lib/SPIRV/SPIRVReader.cpp [hunk 10]
```diff
@@ -4426,3 +4428,32 @@ bool llvm::getSpecConstInfo(std::istream &IS,
...
+const StringSet<> SPIRVToLLVM::BuiltInConstFunc {
```
Comment (AlexeySotkin) [inline lib/SPIRV/SPIRVReader.cpp:4432]:
```suggestion
// clang-format off
const StringSet<> SPIRVToLLVM::BuiltInConstFunc {
```

File: lib/SPIRV/SPIRVReader.cpp [hunk 11]
```diff
@@ -4426,3 +4428,32 @@ bool llvm::getSpecConstInfo(std::istream &IS,
...
+  "sub_sat", "upsample", "popcount", "mad24", "mul24", "degrees", "mix", "radians",
+  "step", "smoothstep", "sign", "cross", "dot", "distance", "length", "normalize",
+  "fast_distance", "fast_length", "fast_normalize", "isequal", "isnotequal",
+  "isgreater", "isgreaterequal", "isless", "islessequal", "islessgreater",
+  "isfinite", "isinf", "isnan", "isnormal", "isordered", "isunordered", "signbit",
+  "any", "all", "bitselect", "select", "shuffle", "shuffle2", "get_image_width",
+  "get_image_height", "get_image_depth", "get_image_channel_data_type",
+  "get_image_channel_order", "get_image_dim", "get_image_array_size",
+  "get_image_array_size", "sub_group_inverse_ballot", "sub_group_ballot_bit_extract",
+  };
```
Comment (AlexeySotkin) [inline lib/SPIRV/SPIRVReader.cpp:4459]:
```suggestion
  };
  // clang-format on
  
```

**Review by AlexeySotkin — APPROVED** [review]

**Pull request merged**